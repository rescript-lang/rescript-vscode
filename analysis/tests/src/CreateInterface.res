// ^int

type r = {name: string, age: int}

let add = (~x, ~y) => x + y

@react.component
let make = (~name) => React.string(name)

@react.component
let otherComponentName = (~name) => React.string(name)

module Mod = {
  @react.component
  let make = (~name) => React.string(name)
}

module type ModTyp = {
  @react.component
  let make: (~name: string) => React.element
}

@module("path") external dirname: string => string = "dirname"

@module("path") @variadic
external join: array<string> => string = "join"

@val
external padLeft: (
  string,
  @unwrap
  [
    | #Str(string)
    | #Int(int)
  ],
) => string = "padLeft"

@inline
let f1 = 10

@inline let f2 = "some string"

@genType
@inline
let f3 = 10

@genType @inline
let f4 = "some string"

@genType @inline let f5 = 5.5

module RFS = {
  @module("fs")
  external readFileSync: (
    ~name: string,
    @string
    [
      | #utf8
      | @as("ascii") #useAscii
    ],
  ) => string = "readFileSync"
}

module Functor = () => {
  @react.component
  let make = () => React.null
}

module type FT = {
  module Functor: (
    X: {
      let a: int
      @react.component
      let make: (~name: string) => React.element
      let b: int
    },
    Y: ModTyp,
  ) =>
  {
    @react.component
    let make: (~name: string) => React.element
  }
}

module NormaList = List
open Belt
module BeltList = List

module type MT2 = ModTyp

module rec RM: ModTyp = D
and D: ModTyp = Mod

module type OptT = {
  @react.component
  let withOpt1: (~x: int=?, ~y: int) => int

  @react.component
  let withOpt2: (~x: int=?, ~y: int) => int

  @react.component
  let withOpt3: (~x: option<int>, ~y: int) => int
}

module Opt = {
  @react.component
  let withOpt1 = (~x=3, ~y) => x + y

  @react.component
  let withOpt2 = (~x=?, ~y) =>
    switch x {
    | None => 0
    | Some(x) => x
    } +
    y

  @react.component
  let withOpt3 = (~x, ~y) =>
    switch x {
    | None => 0
    | Some(x) => x
    } +
    y
}

module Opt2: OptT = Opt
module Opt3 = Opt

module Memo = {
  @react.component
  let make = (~name) => React.string(name)

  let make = React.memo(make)
}
