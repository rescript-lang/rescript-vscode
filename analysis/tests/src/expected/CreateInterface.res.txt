Create Interface tests/src/CreateInterface.res
type r = {name: string, age: int}
let add: (~x: int, ~y: int) => int
@react.component
let make: (~name: string) => React.element
@react.component
let otherComponentName: (~name: string) => React.element
module Mod: {
  @react.component
  let make: (~name: string) => React.element
}
module type ModTyp = {
  @react.component
  let make: (~name: string) => React.element
}
@module("path") external dirname: string => string = "dirname"
@module("path") @variadic
external join: array<string> => string = "join"
@val
external padLeft: (
  string,
  @unwrap
  [
    | #Str(string)
    | #Int(int)
  ],
) => string = "padLeft"
module RFS: {
  @module("fs")
  external readFileSync: (
    ~name: string,
    @string
    [
      | #utf8
      | @as("ascii") #useAscii
    ],
  ) => string = "readFileSync"
}
module Functor: () =>
{
  @react.component
  let make: unit => React.element
}
module type FT = {
  module Functor: (
    X: {
      let a: int
      @react.component
      let make: (~name: string) => React.element
      let b: int
    },
    Y: ModTyp,
  ) =>
  {
    @react.component
    let make: (~name: string) => React.element
  }
}
module NormaList = List
module BeltList = Belt.List
module type MT2 = ModTyp
module rec RM: ModTyp
and D: ModTyp

